File: services/indicator_calculators/rma_calculator.py
from collections.abc import Sequence
from typing import Any

from .utils import calculate_wilder_ma


def calculate_rma(values: Sequence[float | None], period: int) -> dict[str, Any]:
    """
    Calculate RMA (Relative Moving Average / Wilder's Moving Average) indicator.

    Args:
        values: List of values to calculate RMA on (can contain None values)
        period: Period for RMA calculation

    Returns:
        Dictionary with 'rma' as a list of calculated values for each row,
        matching the TypeScript implementation that returns full time series.
        Each list contains the calculated values for corresponding rows in the input data.

    Reference:
        https://www.tradingcode.net/tradingview/relative-moving-average/
    """
    if period <= 0:
        return {"rma": [None] * len(values)}

    # Convert Sequence to list for calculate_wilder_ma
    values_list = list(values)
    # Use Wilder's MA from utils
    rma_values = calculate_wilder_ma(values_list, period)

    return {"rma": rma_values}

File: backend/src/services/indicator-calculators/rma-calculator.ts
// src/services/indicator-calculators/rma-calculator.ts
// Translated from: services/indicator_calculators/rma_calculator.py

import { calculateWilderMa } from './utils';

export interface RmaResult {
  rma: (number | null)[];
}

/**
 * Calculate RMA (Relative Moving Average / Wilder's Moving Average) indicator.
 *
 * @param values - List of values to calculate RMA on (can contain null values)
 * @param period - Period for RMA calculation
 * @returns Dictionary with 'rma' as a list of calculated values for each row
 *
 * Reference: https://www.tradingcode.net/tradingview/relative-moving-average/
 */
export function calculateRma(
  values: readonly (number | null)[],
  period: number
): RmaResult {
  if (period <= 0) {
    return { rma: new Array(values.length).fill(null) };
  }

  // Use Wilder's MA from utils
  const rmaValues = calculateWilderMa([...values], period);

  return { rma: rmaValues };
}
